### PR TITLE
rebased onto upstreams stable v239

### PR DESCRIPTION
**Pull request by andir**
This updates our v239 branch to the latest upstream stable v239 version.

There are a few fixes in here. I initially was just aiming for fixing the DHCPv6 out out bounds access (CVE-2018-15688) but now we might have a bunch more.

I ran a bunch of VM tests (installers and others) from master against this version without any issues. I rebuild my desktop (18.03, local backport) with `config.systemd.package` set to the updated version. It booted fine and I am typing on it right now..

The important thing to change in nixpkgs besides just updating the hashes is to also disable `dns-over-tls` since that requires gnutls. We can also pull in gnutls but I decided against introducing that as dependency when testing.

```patch
diff --git a/pkgs/os-specific/linux/systemd/default.nix b/pkgs/os-specific/linux/systemd/default.nix
--- a/pkgs/os-specific/linux/systemd/default.nix
+++ b/pkgs/os-specific/linux/systemd/default.nix
@@ -91,6 +91,7 @@ in stdenv.mkDerivation rec {
     "-Dsulogin-path=${utillinux}/bin/sulogin"
     "-Dmount-path=${utillinux}/bin/mount"
     "-Dumount-path=${utillinux}/bin/umount"
+    "-Ddns-over-tls=false"
   ];

```

cc @Mic92 @fpletz

### CODE REVIEW

File: src/network/netdev/wireguard.c [hunk 1]
```diff
@@ -108,7 +108,7 @@ static int set_wireguard_interface(NetDev *netdev) {
                                 if (r < 0)
                                         break;
 
-                                r = sd_netlink_message_append_u32(message, WGPEER_A_PERSISTENT_KEEPALIVE_INTERVAL, peer->persistent_keepalive_interval);
+                                r = sd_netlink_message_append_u16(message, WGPEER_A_PERSISTENT_KEEPALIVE_INTERVAL, peer->persistent_keepalive_interval);
```
Comment (Mic92) [inline src/network/netdev/wireguard.c:111]: oh snap :)

File: units/systemd-networkd.service.in [hunk 2]
```diff
@@ -25,9 +25,9 @@ RestartSec=0
 ExecStart=!!@rootlibexecdir@/systemd-networkd
 WatchdogSec=3min
 User=systemd-network
-DynamicUser=yes
```
Comment (Mic92) [inline units/systemd-networkd.service.in:28]: We have a systemd-network user right?

**Comment by Mic92** [comment]
Wow these are quite a few updates. I would suggest we have a grace period before we backport this.

**Comment by grahamc** [comment]
Trying this via

```nix
{
  systemd.package = (import (pkgs.fetchFromGitHub {
    owner = "andir";
    repo = "nixpkgs";
    rev = "5fd9f68276f7b91f2b2d2336c2d1cc5c5f232adc";
    sha256 = "1bj3xql0fw0a1qbxvvl02lsfpxhacg844p1qswp1y5vwp98xn5la";
    }) {}).systemd;
}
```

works fine

File: units/systemd-networkd.service.in (hunk 2 above)
Comment (andir) [reply units/systemd-networkd.service.in:28]: I haven't seen were we define it. My guess right now is that systemd is creating that transparently for us? On the three machines I am running this it looks fine.

**Comment by andir** [comment]
Because of the severity of some of the improvements (RCE, …) I'd argue that the grace period shouldn't be too long.

File: units/systemd-networkd.service.in (hunk 2 above)
Comment (Mic92) [reply units/systemd-networkd.service.in:28]: Its in `modules/system/boot/systemd.nix`

**Comment by vcunat** [comment]
Stable branches: we might better just apply important patches, if it's not too complex.  Systemd still tends to do relatively intrusive changes &ndash; now the tests have caught https://github.com/NixOS/systemd/commit/6789dca0a26df0 but I'm more afraid of breaking changes _not_ caught by tests just getting to stable.

**Comment by andir** [comment]
I agree.. Right now I really like the idea of just reverting the changes in
nixokgs and working on a cleaner approach..

**Comment by vcunat** [comment]
Do they collect links to such important fixes somewhere?  I don't see the security fix in `NEWS` on their master, so that doesn't seem a good source.

**Comment by vcunat** [comment]
Hmm, they publish ["stable" branches](https://github.com/systemd/systemd-stable/tree/v239-stable), but I haven't read any statement about them.  There still seems to be relatively lots of changed code lines between v239 and current v239-stable, so I don't expect we need to have even nixpkgs master on anything more unstable.

**Comment by andir** [comment]
Those `stable` branches is exactly what I used there.. They are very intrusive as it turned out. A lot more then what I did expect. Even after reading through all of the commits I thought it would be mostly *additional* flags and switches and not breaking changes as with the hibernation change.

The MLs are also not very noisy in regards to backports fixes... :/

**Comment by vcunat** [comment]
Oh, I didn't realize.  I see they have roughly 10% of line changes, compared to master, so they _are_ much smaller but they still change a couple thousand lines of C code :-)

**Comment by andir** [comment]
I started iterating on the bugs that we found and (potentially haven't found) with #25. I think the patchset there can be made obsolete in favor of what I pushed to https://github.com/andir/systemd/commits/nixos-v239-test.

The question for the intermediate time is: Should be risk using one of the fixed versions (#25 or https://github.com/andir/systemd/commits/nixos-v239-test) or revert nixpkgs for now and try to come up with a minimal fix for all the issues that we can identify?

Given that the channels of 18.09 and master are blocked and we haven't been able to bump them since before firefox63, chromium70, … were merged I tend towards reverting for now.

Even after reverting the current (failed) attempt on 18.09 & master we can try to run more tests against one of the proposed patch sets.

The changes done will probably be required as soon as we decide to port systemd v240 to NixOS anyway...

**Comment by vcunat** [comment]
One way, for our stable releases at least, is to feed off some other packagers that we trust to get (all) important fixes.  Well, we can also add some they don't have, if we notice, but doing all the work might be much work for us in systemd case.  For example, [Debian's stretch-backports](https://salsa.debian.org/systemd-team/systemd/commits/stretch-backports/debian/patches) is on 239 just as our 18.09.

**Comment by vcunat** [comment]
OK, let me test that approach for 18.09.  Local rebuilds in progress.

**Comment by vcunat** [comment]
Tests were running fine, including hibernation, so I pushed that to 18.09 to unblock it.

Shall I do the same for master, or do you have other plans?  (e.g. keep the version and quickly fixup any regressions)

**Comment by andir** [comment]
Feel free to do the same there. I think it is the better approach for now.

**Pull request merged**